src/components/Modals/components/ModalComponents.js

```diff
@@ -0,0 +1,70 @@
+/**
```

Comment:
@motine @MatthewBerglund: I took Tom's suggestion to refactor the modals in a more re-usable way and avoiding to pollute the DOM. I came up with a dynamic and customisable Modal&Components file which includes all you need to make your own modal (including docs).

Just to explain this approach: first I tried to make ONE modal which could receive props but in order to make it dynamic the list of necessary props was passing the double-digit (e.g. `modalId`, `headerTitle`, `bodyTitle`, `expense`, `setModalOpen`, `buttonLabel1`, `buttonLabel2`, `buttonAction1`, `buttonAction2`, `buttonColour1`, and many more). Moreover, several if statements or even switch statements were necessary to render dynamically. 

I have then rather decided to use the "_sub-classes_" like approach which allows us to use basic components (Modal, header, body, footer) and pass the Bootstrap HTML we want as `props.children` and the React state logic in a more visual and flexible way. 

Let me know what you think about it

Reply:
thanks for the refactor. this is already much better.

i like the approach to have individual components for different modals.

i see two ideas to improve the design further:

**Option 1**

you can pass multiple slots to a component. here the "API" of the modal can be misused by the developer (e.g. forget the footer or screw up the order of things). also, the components like ModalFooter do not make sense on their own. here an idea with two slots (not exact code, just an initial idea):

```js
<Modal
  title="Confirm expense deletion"
  footer={
    <button type="button" className="btn btn-secondary" onClick={() => cancelClicked()} >I'm not sure</button>
    <button type="button" className="btn btn-danger" ref={callForAction} onClick={() => okClicked()}>Yes delete!!</button>
  }
  okCallback={...}
  cancelCallback={...}
> <!-- the modal id can be autogenerated by the Modal component-->

  <!-- the actual body goes here-->
  <h6 className="mb-3">
    Please confirm you want to delete this expense:
  </h6>
  {expense && (
    <dl className="row">
      <dt className="col-3">Date:</dt>
      <dd className="col-9">{getUKFormattedDate(date)}</dd>
      <dt className="col-3">Title:</dt>
      <dd className="col-9">{title}</dd>
      <dt className="col-3">Amount:</dt>
      <dd className="col-9">{getUKFormattedEuros(amount)}</dd>
    </dl>
  )}
</Modal>
```

**Option 2**

Actual subclasses:

```js
class SomeModal extends Modal {
  title() {
    return "This is the modal title";
  }

  body() {
    return <>
      <h6 className="mb-3">
        Please confirm you want to delete this expense:
      </h6>
      <dl className="row">
        <dt className="col-3">Date:</dt>
        <dd className="col-9">{getUKFormattedDate(date)}</dd>
        <dt className="col-3">Title:</dt>
        <dd className="col-9">{title}</dd>
        <dt className="col-3">Amount:</dt>
        <dd className="col-9">{getUKFormattedEuros(amount)}</dd>
      </dl>
    </>
  }

  footer() {
    return <>
      <button type="button" className="btn btn-secondary" onClick={() => cancelClicked()} >I'm not sure</button>
      <button type="button" className="btn btn-danger" ref={callForAction} onClick={() => okClicked()}>Yes delete!!</button>
    </>
  }
}
```

Depending on how motivated you are on working on this again, we can just leave it as it is or do it in another pull request. i think reworking this is a great learning opportunity to see different versions in action.

Reply:
@motine as mentioned in the chat, I opted for the first option and passed a couple of props... now `Ref` and `cancelCallback` methods are incapsulated in the `Modal` components and cannot be _misused_. Specific data, props and methods can be passed from `App.js`. In case we would need further customisation, `Modal` and `modalProps` can be extended. Let me know what you think about it!

Reply:
nice work. I think this is much better now :)